File: data_structures/trie.py
class Trie:
    class node:
        def __init__(self, symb = None, parent = None):
            self.count = 0
            self.parent = parent
            self.symb = symb

            # symbol -> node
            self.child = dict()
        
    def __init__(self):
        self.root = Trie.node()

    def add_word(self, w):
        curr = self.root
        for c in w:
            if c not in curr.child:
                curr.child[c] = Trie.node(c, curr)
            curr = curr.child[c]
        curr.count += 1

    def dfs(self, v = None):
        if(v is None):
            v = self.root

        print('symb:', v.symb);print('count:', v.count)

        for symb, u in v.child.items():
            self.dfs(u)

if __name__ == "__main__":
    
    t = Trie()
    t.add_word('hello!')
    t.add_word('world.')
    t.add_word('hell')

    t.dfs()

File: data_structures/trie.cpp
#include <bits/stdc++.h>

using namespace std;
#define mp make_pair
#define all(x) (x).begin(), (x).end()
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int,int> pii;
typedef long long ll;

class trie{
public:
    class node{
    public:
        char c;
        node* p;
        int count;
        map<char, node*> child;
        node(char c = -1, node* p = nullptr) : c(c) , p(p) {
            count = 0;
        }
        ~node(){
            for(auto u : child){ delete u.second;}
        }
    };
    node *root;

    trie(){ root = new node(); }

    ~trie(){ delete root; }

    void insert(vector<char> word){
        auto curr = root;
        for(char ch : word){
            auto u = curr->child.find(ch);
            if(u == curr->child.end()){
                node *uu = new node(ch, curr);
                curr->child[ch] = uu;
                curr = uu;
            }else{
                curr = (*u).second;
            }
        }
        curr->count += 1;
    }
    void insert(string s){
        insert(vector<char>(s.begin(), s.end()));
    }

    void dfs(node *v = nullptr){
        if(v == nullptr)
            v = root;
        
        cout << "symb = " << v->c << " count = " << v->count << endl;

        for(auto u : v->child){
            node *uu = u.second;
            dfs(uu);
        }
    }
};

int main(){
    trie t;
    t.insert("hello!");
    t.insert("world.");
    t.insert("hell");
    t.dfs();
    return 0;
}